First draft of reference docs on durable state (#30387)
akka/akka: akka-docs/src/main/paradox/typed/persistence-style-durable-state.md

# Style Guide 

@@@ div { .group-scala }
## Command handlers in the state

> We can take the previous bank account example one step further by handling the commands in the state too.

Reviewer: ```suggestion
We can take the previous bank account example one step further by handling the commands within the state as well.
```

Author: done ..

Scala
:  @@snip [AccountExampleWithCommandHandlersInDurableState.scala](/akka-cluster-sharding-typed/src/test/scala/docs/akka/cluster/sharding/typed/AccountExampleWithCommandHandlersInDurableState.scala) { #account-entity }

> Notice how the command handler is delegating to `applyCommand` in the `Account` (state), which is implemented

Reviewer: ```suggestion
Take note of how the command handler is delegating to `applyCommand` in the `Account` (state), which is implemented
```

Author: done ..

in the concrete `EmptyAccount`, `OpenedAccount`, and `ClosedAccount`.

@@@

## Optional initial state

> Sometimes it's not desirable to use a separate state class for the empty initial state, but rather treat that as
> there is no state yet.

Reviewer: ```suggestion
Sometimes it's not desirable to use a separate state class for the empty initial state, but rather act as if
there is no state yet.
```

Author: done ..

> @java[`null` can then be used as the `emptyState`, but be aware of that the `state` parameter

Reviewer 2: ```suggestion
@java[You can use `null` can as the `emptyState`, but be aware of that the `state` parameter
```

Reviewer: I think the second 'can' is unneeded - (after 'null')
@java[You can use `null` as the `emptyState`, but be aware of that the `state` parameter

Author: done ..

> will then be `null` for the first commands until the first non-null state has been persisted 

Reviewer 2: ```suggestion
will be `null` until the first non-null state has been persisted 
```

Author: done ..

> It's possible to use `Optional` instead of `null` but that results in rather much boilerplate

Reviewer 2: ```suggestion
It's possible to use `Optional` instead of `null`, but that requires extra boilerplate
```

Author: done ..

> to unwrap the `Optional` state parameter and therefore `null` is preferred. The following example

Reviewer 2: ```suggestion
to unwrap the `Optional` state parameter. Therefore, use of  `null` is simpler. The following example
```

Author: done ..

> to unwrap the `Optional` state parameter and therefore `null` is probably preferred. The following example

Reviewer 3: ```suggestion
to unwrap the `Optional` state parameter and therefore `null` is preferred. The following example
```

Author: done ..

illustrates using `null` as the `emptyState`.]

> @scala[`Option[State]` can be used as the state type and `None` as the `emptyState`. Pattern matching
> is then used in command handlers at the outer layer before delegating to the state or other methods.]

Reviewer: ```suggestion
@scala[`Option[State]` can be used as the state type and `None` as the `emptyState`. Then pattern matching
is used in command handlers at the outer layer, before delegating to the state or other methods.]
```

Author: done ..
